pkg/apis/sources/v1beta1/ping_conversion.go

```diff
@@ -18,17 +18,130 @@ package v1beta1
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+
+	"knative.dev/eventing/pkg/apis/sources/v1beta2"
 	"knative.dev/pkg/apis"
 )
 
+const (
+	// V1B1SpecAnnotationKey is used to indicate that a v1beta2 object is converted from v1beta1
+	// also it can be used to downgrade such object to v1beta1
+	V1B1SpecAnnotationKey = "pingsources.sources.knative.dev/v1beta1-spec"
+
+	// V1B2SpecAnnotationKey is used to indicate that a v1beta1 object is converted from v1beta2
+	// also it can be used to convert the v1beta1 object back to v1beta2, considering that v1beta2 introduces more features.
+	V1B2SpecAnnotationKey = "pingsources.sources.knative.dev/v1beta2-spec"
+)
+
+type message struct {
+	Body string `json:"body"`
+}
+
+func makeMessage(body string) ([]byte, error) {
+	// try to marshal the body into an interface.
+	var objmap map[string]*json.RawMessage
+	if err := json.Unmarshal([]byte(body), &objmap); err != nil {
+		// Default to a wrapped message.
+		return json.Marshal(message{Body: body})
+	}
+	return json.Marshal(objmap)
+}
+
 // ConvertTo implements apis.Convertible
-func (source *PingSource) ConvertTo(ctx context.Context, sink apis.Convertible) error {
-	return fmt.Errorf("v1beta1 is the highest known version, got: %T", sink)
+// Converts source from v1beta1.PingSource into a higher version.
+func (source *PingSource) ConvertTo(ctx context.Context, obj apis.Convertible) error {
+	switch sink := obj.(type) {
+	case *v1beta2.PingSource:
+		sink.ObjectMeta = source.ObjectMeta
+		sink.Status = v1beta2.PingSourceStatus{
+			SourceStatus: source.Status.SourceStatus,
+		}
+
+		// deep copy annotations to avoid mutation on source.ObjectMeta.Annotations
+		annotations := make(map[string]string)
+		for key, value := range source.GetAnnotations() {
+			annotations[key] = value
+		}
+
+		// try to unmarshal v1beta2.PingSource.Spec from V1B2SpecAnnotationKey
+		v1beta2Spec, ok := annotations[V1B2SpecAnnotationKey]
+		if ok {
+			if err := json.Unmarshal([]byte(v1beta2Spec), &sink.Spec); err != nil {
+				ok = false
+			}
+			// we don't need this annotation in a v1beta2.PingSource object
+			delete(annotations, V1B2SpecAnnotationKey)
+		}
+
+		// cannot unmarshal, do a normal conversion
+		if !ok {
```

Comment:
nit: maybe just else?

Reply:
nvm, I saw you are setting ok false inside the other if

Reply:
this !ok can happen if the annotation doesn't exist or there was a problem with unmarshalling? Is that right?
If that is the case, then you should probably update the comment.